### PR TITLE
[Bug:] `azurerm_cosmosdb_account` - add default values for the `consistency_policy` fields and fix regression in the `backup` code block

### PR DESCRIPTION
**Pull request by WodansSon**

### CODE REVIEW

File: internal/services/cosmos/cosmosdb_account_resource.go [hunk 1]
```diff
@@ -1907,28 +1909,37 @@ func expandCosmosdbAccountBackup(input []interface{}, backupHasChange bool, crea
...
-			return nil, fmt.Errorf("`storage_redundancy` can not be set when `type` in `backup` is `Continuous`")
+			return nil, fmt.Errorf("`storage_redundancy` cannot be defined when the `backup.type` is set to %q", string(cosmosdb.BackupPolicyTypeContinuous))
 		}
+
 		return cosmosdb.ContinuousModeBackupPolicy{}, nil
 
 	case string(cosmosdb.BackupPolicyTypePeriodic):
 		if createMode != "" {
-			return nil, fmt.Errorf("`create_mode` only works when `backup.type` is `Continuous`")
+			return nil, fmt.Errorf("`create_mode` can only be defined when the `backup.type` is set to %q, got %q", string(cosmosdb.BackupPolicyTypeContinuous), string(cosmosdb.BackupPolicyTypePeriodic))
```
Comment (manicminer) [inline internal/services/cosmos/cosmosdb_account_resource.go:1927]:
```suggestion
			return nil, fmt.Errorf("`create_mode` can only be defined when the `backup.type` is set to %q, got %q", cosmosdb.BackupPolicyTypeContinuous, cosmosdb.BackupPolicyTypePeriodic)
```

File: internal/services/cosmos/cosmosdb_account_resource_test.go [hunk 2]
```diff
@@ -992,7 +992,27 @@ func TestAccCosmosDBAccount_identity(t *testing.T) {
 	})
 }
 
-func TestAccCosmosDBAccount_backup(t *testing.T) {
+func TestAccCosmosDBAccount_icmRegression(t *testing.T) {
+	// Regression test for the Microsoft IcM 466814988...
```
Comment (manicminer) [inline internal/services/cosmos/cosmosdb_account_resource_test.go:996]: Could this test have a more descriptive name? We can also leave out the IcM number comment as these are MSFT-internal.

File: website/docs/r/cosmosdb_account.html.markdown [hunk 3]
```diff
@@ -194,7 +194,7 @@ The `consistency_policy` block Configures the database consistency and supports
   
 * `max_staleness_prefix` - (Optional) When used with the Bounded Staleness consistency level, this value represents the number of stale requests tolerated. The accepted range for this value is `10` – `2147483647`. Defaults to `100`. Required when `consistency_level` is set to `BoundedStaleness`.
 
-~> **Note:** `max_interval_in_seconds` and `max_staleness_prefix` can only be set to custom values when `consistency_level` is set to `BoundedStaleness` - otherwise they will return the default values shown above.
+~> **NOTE:** `max_interval_in_seconds` and `max_staleness_prefix` can only be set to values other than default when the `consistency_level` is set to `BoundedStaleness`.
```
Comment (manicminer) [inline website/docs/r/cosmosdb_account.html.markdown:197]: (capitalization here is legacy pattern)

```suggestion
~> **Note** `max_interval_in_seconds` and `max_staleness_prefix` can only be set to values other than default when the `consistency_level` is set to `BoundedStaleness`.
```

File: website/docs/r/cosmosdb_account.html.markdown [hunk 4]
```diff
@@ -241,13 +241,17 @@ A `capacity` block supports the following:
 
 A `backup` block supports the following:
 
-* `type` - (Required) The type of the `backup`. Possible values are `Continuous` and `Periodic`. Migration of `Periodic` to `Continuous` is one-way, changing `Continuous` to `Periodic` forces a new resource to be created.
+* `type` - (Required) The type of the `backup`. Possible values are `Continuous` and `Periodic`. 
 
-* `interval_in_minutes` - (Optional) The interval in minutes between two backups. This is configurable only when `type` is `Periodic`. Possible values are between 60 and 1440.
+~> **NOTE:** Migration of `Periodic` to `Continuous` is one-way, changing `Continuous` to `Periodic` forces a new resource to be created.
```
Comment (manicminer) [inline website/docs/r/cosmosdb_account.html.markdown:246]:
```suggestion
~> **Note** Migration of `Periodic` to `Continuous` is one-way, changing `Continuous` to `Periodic` forces a new resource to be created.
```

File: website/docs/r/cosmosdb_account.html.markdown [hunk 5]
```diff
@@ -241,13 +241,17 @@ A `capacity` block supports the following:
...
 
-* `retention_in_hours` - (Optional) The time in hours that each backup is retained. This is configurable only when `type` is `Periodic`. Possible values are between 8 and 720.
+* `interval_in_minutes` - (Optional) The interval in minutes between two backups. Possible values are between 60 and 1440. Defaults to `240`.
 
-* `storage_redundancy` - (Optional) The storage redundancy is used to indicate the type of backup residency. This is configurable only when `type` is `Periodic`. Possible values are `Geo`, `Local` and `Zone`.
+* `retention_in_hours` - (Optional) The time in hours that each backup is retained. Possible values are between 8 and 720. Defaults to `8`.
+
+* `storage_redundancy` - (Optional) The storage redundancy is used to indicate the type of backup residency. Possible values are `Geo`, `Local` and `Zone`. Defaults to `Geo`.
+
+~> **NOTE:** You can only configure `interval_in_minutes`, `retention_in_hours` and `storage_redundancy` when the `type` field is set to `Periodic`.
```
Comment (manicminer) [inline website/docs/r/cosmosdb_account.html.markdown:254]:
```suggestion
~> **Note** You can only configure `interval_in_minutes`, `retention_in_hours` and `storage_redundancy` when the `type` field is set to `Periodic`.
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hey @WodansSon, thanks for the PR. This mostly LGTM, just a few minor linting/naming comments if you could take a look at those. Thanks!

File: website/docs/r/cosmosdb_account.html.markdown (hunk 5 above)
Comment (WodansSon) [reply website/docs/r/cosmosdb_account.html.markdown:254]: Fixed.

File: website/docs/r/cosmosdb_account.html.markdown (hunk 3 above)
Comment (WodansSon) [reply website/docs/r/cosmosdb_account.html.markdown:197]: Fixed.

File: internal/services/cosmos/cosmosdb_account_resource_test.go (hunk 2 above)
Comment (WodansSon) [reply internal/services/cosmos/cosmosdb_account_resource_test.go:996]: Fixed.

File: internal/services/cosmos/cosmosdb_account_resource.go (hunk 1 above)
Comment (WodansSon) [reply internal/services/cosmos/cosmosdb_account_resource.go:1927]: Fixed.

File: internal/services/cosmos/cosmosdb_account_resource.go [hunk 6]
```diff
@@ -1907,28 +1909,37 @@ func expandCosmosdbAccountBackup(input []interface{}, backupHasChange bool, crea
 	switch attr["type"].(string) {
 	case string(cosmosdb.BackupPolicyTypeContinuous):
 		if v := attr["interval_in_minutes"].(int); v != 0 && !backupHasChange {
-			return nil, fmt.Errorf("`interval_in_minutes` can not be set when `type` in `backup` is `Continuous`")
+			return nil, fmt.Errorf("`interval_in_minutes` cannot be defined when the `backup.type` is set to %q", string(cosmosdb.BackupPolicyTypeContinuous))
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply internal/services/cosmos/cosmosdb_account_resource.go:1912]: Fixed.

**Review by katbyte — APPROVED** [review]
LGTM 🛻

**Comment by WodansSon** [comment]
Failing test cases are not related to this PR:

image

**Pull request merged**